models/autoencoders/discreteVAE.py

```diff
@@ -11,353 +11,344 @@
+            undifferentiated_component=undifferentiated_component.detach()
+        
+        #final KLD gradient wrt phi
+        kld_per_sample = torch.sum(undifferentiated_component * posterior_probs, dim=1)
+
+        return kld_per_sample
+
+    def kl_divergence(self, posterior_distribution , posterior_samples):
+        logger.debug("kl_divergence")
+        #posterior_distribution: distribution with logits from each hierarchy level/layer
+        #posterior_samples: reparameterised output of posterior_distribution
+        if len(posterior_distribution)>1 and self.training:
+
+            logit_list=[]
+            samples_marginalised=[]
+            for lvl in range(len(posterior_distribution)):
+
+                current_post_dist=posterior_distribution[lvl]
+                current_post_samples=posterior_samples[lvl]
+
+                logits=torch.clamp(current_post_dist.logits,min=-88,max=88)
+                logit_list.append(logits)
+        
+                #TODO this step is not clear anymore: where was this motivated
+                #in the paper? 
+                if lvl==len(posterior_distribution)-1:
+                    samples_marginalised.append(torch.sigmoid(logits))
+                else:
+                    zero_mask=torch.zeros(current_post_samples.size())
+                    one_mask=torch.ones(current_post_samples.size())
+                    post_sample_marginalised=torch.where(current_post_samples>0.0,one_mask,zero_mask)
+                    samples_marginalised.append(post_sample_marginalised)
+
+            logits_concat=torch.cat(logit_list,1)
+            samples_marginalised_concat=torch.cat(samples_marginalised,1)
+
+            kl_div_posterior_distribution=self.kl_div_posterior_gradient(
+                posterior_logits=logits_concat,
+                posterior_binary_samples=samples_marginalised_concat)
+                
+            kl_div_prior=self.kl_div_prior_gradient(
+                posterior_logits=logits_concat,
+                posterior_binary_samples=samples_marginalised_concat)  #DVAE Eq11 - gradient of prior   
+            kld=kl_div_prior+kl_div_posterior_distribution 
+            return kld
+        else: # either this posterior only has one latent layer or we are not looking at training
+            # #this posterior is not hierarchical - a closed analytical form for the KLD term can be constructed
+            # #the mean-field solution (n_latent_hierarchy_lvls == 1) reduces to log_ratio = 0.
+            # logger.debug("kld for evaluation/training of one layer posterior")
+            return 0
+
+    def generate_samples(self):
+        prior_samples=[]
+        #how many samples (i.e. digits) to look at
+        for i in range(0,self._config.n_generate_samples):
+            prior_sample = self.sampler.get_samples(
+                approx_post_samples=[], #empty list, as we want to generate samples from random numbers
```

Comment:
All the changes in this file are re-tabbing of the indentation.

This file has not been changed aside from this line.